AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "MSKeychainAuthTokenStorage.h"
+#import "MSAuthTokenInfo.h"
+#import "MSIdentityConstants.h"
+#import "MSIdentityPrivate.h"
+#import "MSKeychainUtil.h"
+#import "MSLogger.h"
+#import "MSUtility.h"
+
+@implementation MSKeychainAuthTokenStorage
+
+- (nullable NSString *)retrieveAuthToken {
+  NSString *authToken = [MSKeychainUtil stringForKey:kMSIdentityAuthTokenKey];
+  if (authToken) {
+    MSLogDebug([MSIdentity logTag], @"Retrieved auth token from keychain.");
+  } else {
+    MSLogWarning([MSIdentity logTag], @"Failed to retrieve auth token from keychain or none was found.");
+  }
+  return authToken;
+}
+
+- (nullable NSString *)retrieveAccountId {
+  return [MS_USER_DEFAULTS objectForKey:kMSIdentityMSALAccountHomeAccountKey];
+}
+
+- (MSAuthTokenInfo *)oldestAuthToken {
+
+  // Read token array from storage.
+  NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+  if (tokenArray.count == 0) {
+    return nil;
+  }
+
+  MSAuthTokenInfo *authTokenInfo = tokenArray.firstObject;
+  NSDate *nextChangeTime = tokenArray.count > 1 ? tokenArray[1].startTime : nil;
+  if ([authTokenInfo.endTime laterDate:nextChangeTime]) {
+    return [[MSAuthTokenInfo alloc] initWithAuthToken:authTokenInfo.authToken
+                                         andStartTime:authTokenInfo.startTime
+                                           andEndTime:nextChangeTime];
+  }
+  return authTokenInfo;
+}
+
+- (void)saveAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+    if (tokenArray.count == 0) {
+
+      // Add nil token if the entire story is empty.
+      [tokenArray addObject:[MSAuthTokenInfo new]];
+    }
+    if (![tokenArray.lastObject.authToken isEqual:authToken]) {
+
+      // If new token differs from the last token of array - add it to array.
+      MSAuthTokenInfo *newAuthToken = [[MSAuthTokenInfo alloc] initWithAuthToken:authToken andStartTime:[NSDate date] andEndTime:expiresOn];
+      [tokenArray addObject:newAuthToken];
+    }
+
+    // Cap array size at max available size const (deleting from beginning).
+    if ([tokenArray count] > kMSIdentityMaxAuthTokenArraySize) {
+      [tokenArray removeObjectAtIndex:0];
```

Comment:
Can you explain why at index 0? Have you tested how it works in iOS? I'm just not sure if the array items will move or the position will stay empty.. Maybe we should store oldest tokens at the end of array?

Reply:
Filling gaps works as subtracting 1 from every index beyond (according to apple's [dev portal](https://developer.apple.com/documentation/foundation/nsmutablearray/1416616-removeobjectatindex?language=objc)) so no, they will not stay empty.
As for why at 0 - we need to delete oldest entry to maintain array size, and since those are in the beginning (newest are written to the end), we delete element at 0.